share/doc/wake/logging.md

```diff
@@ -0,0 +1,71 @@
+# Logging
+
+This article gives a high level overview of how wake logging works.
+Check out [print.wake](https://github.com/sifive/wake/blob/master/share/wake/lib/core/print.wake)
+for commented code.
+
+## How Logging works on Wake Execution commands
+
+Every `Job` run by wake will have its own stdout and stderr,
+and when the underlying process prints to stdout or stderr,
+it will go the `Job`-specific stdout and stderr,
+not the stdout and stderr of the `wake` process itself.
+
+When writing your wake code,
+you can control whether these `Job`-specific stdout and stderr are also printed to `wake`'s own stdout,
+stderr, or other arbitrary file descriptors.
+These can be controlled with `setPlanStderr` and `setPlanStdout` before you `runJob`.
+We call this which _logger_ each goes to (wake code often refers to a _logger_ as a `logLevel`).
+
+Regardless of to which logger you direct the `Job`'s `Stderr` and `Stdout`,
+`wake` will store all the text in its database,
+discarding the information about the logger it was destined for at the time it was stored.
+It will however know whether it came from the `Job`'s stderr/stdout.
+
+So what's the difference between the different loggers?
+What you see on your console when you run wake.
+It controls *what* you see (depending on the command line flags) and what it looks like (what color it is).
+It also controls which *actual* stream it is going to coming out of `wake`
+(stderr or stdout, or other file descriptors 3-5).
+
+There are a number of built in loggers in wake which all start with `log...`.
+You can also make your own logger using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as,
+or (more commonly) to grab output from a `Job` for consuming within your wake program,
+or redirecting outside of wake using shell redirect.
+
+Unless overridden with `--fd:2=...` the *only* thing that ever appears on stderr coming out of wake is `logError`,
+and it always appears there.
+
+For stdout, the following table should be read top to bottom considering the command line arguments you've passed to wake.
+If you match a line in the table, stop!
+This is what will appear on the console as you watch wake execute, and what is sent to wake's stdout:
+
+|Flag               | logDebug | logInfo | logEcho | logReport | logWarning | logError | logNever | mkLogLevel "foo" Green|
```

Comment:
>  but it's probably a good idea to clarify that logEcho is the command run

I dont know what you mean, `logEcho` and `logInfo` are both "log level"s of the same sort of shape? One has name "info" and one has name "echo" and they have different colors.

I think what you are saying is that there are OTHER things, not just running-job-stderr/stdout and println/print that show up on the console, and these are:

* The command run (will always get sent to the `logEcho` logger)

Not sure what you mean by 

> logInfo is the internal Job's stdout.

What is "the internal Job"?

And then I want to flesh out what @richardxia said about the results getting sent to some particular logger.

> f I do a printlnLevel logError does that get printed to Wake's stdout just formatted as if it were an error?

That is my understanding

Reply:
> > logInfo is the internal Job's stdout.
> 
> What is "the internal Job"?

Sorry, the command that is being shelled out for -- so in `makePlan "" Nil "echo Hello world!"` the "Hello world!" is output on `logInfo` by default.  That mapping is not quite obvious from the "Info" name, so it might be worth an explicit mention.

> The command run (will always get sent to the logEcho logger)

I said this more clearly below with the "how noisily the command line/label" comment.  The full `"echo \\"Hello world!\\""` command gets printed to `logEcho` which may or may not be what people expect from *that* name, depending on how familiar they are with things like Batch's `@echo off`/Bash's `set +x`.

Reply:
> the "Hello world!" is output on logInfo by default.

I believe that is equivalent to saying "if you don't override setPlanStdout for a Job, it goes to `logInfo`". (And probably I should say "if you don't override setPlanStderr for a Job, it goes to `logError`" (assuming that is true?)

Reply:
It is true, and if you're already saying that, then great!  I probably just didn't see it, or saw it after writing the comment and forgot to go back to this.

Reply:
no I am not saying that yet, I am going to break this out into a more explicit section based on this feedback

Reply:
ok i think i broke this out more clearly into the different "kinds" of things we are talking about

Reply:
Sorry, what was the rationale for not mentioning the colors separately from the command-line options? I still feel like it's a bit confusing to see orthogonal data in one table like this.